jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/CallbackCompletionHandler.java

```diff
@@ -0,0 +1,874 @@
+        {
+            private final Runnable callback;
+            private final ThreadInfo demandThreadInfo;
+            // Tri-state: null -> no demand, thread -> running demand, this -> done running demand.
+            private volatile Object demandRunner;
+
+            private DemandCallback(Runnable callback)
+            {
+                this.callback = callback;
+                this.demandThreadInfo = new ThreadInfo(Thread.currentThread());
+            }
+
+            @Override
+            public void run()
+            {
+                demandRunner = Thread.currentThread();
+                Scheduler.Task task = getComponents().getScheduler().schedule(this::expired, getDemandCallbackTimeout());
+                try
+                {
+                    callback.run();
+                }
+                finally
+                {
+                    demandRunner = this;
+                    stateInfo.demandCallback = null;
+                    task.cancel();
+                }
+            }
+
+            private void expired()
+            {
+                Object demandRunner = this.demandRunner;
+                // If expired() lost the race with run(), return.
+                if (demandRunner == this)
+                    return;
+
+                // Avoid clash with Request.LOG.
+                Logger log = CallbackCompletionHandler.LOG;
+                if (log.isDebugEnabled())
+                    log.debug("demand callback blocked more than {} for {}", getDemandCallbackTimeout(), getWrapped());
+
+                // The thread info of the thread that blocks.
+                ThreadInfo runThreadInfo = new ThreadInfo((Thread)demandRunner);
+                notifyDemandCallbackBlocked(getWrapped(), demandThreadInfo, runThreadInfo);
+            }
+
+            @Override
+            public void dump(Appendable out, String indent) throws IOException
+            {
+                Object demandRunner = this.demandRunner;
+                if (demandRunner instanceof Thread runThread)
+                {
+                    out.append("demand: running\n");
+                    out.append(indent).append(new ThreadInfo(runThread).toString(indent));
+                }
+                else
+                {
+                    out.append("demand: %s\n".formatted(demandRunner == null ? "pending" : "none"));
+                }
+            }
+        }
+    }
+
+    private class ResponseWrapper extends Response.Wrapper
+    {
+        private final StateInfo stateInfo;
+
+        private ResponseWrapper(StateInfo stateInfo, Response wrapped)
+        {
+            super(stateInfo.request, wrapped);
+            this.stateInfo = stateInfo;
+        }
+
+        @Override
+        public void write(boolean last, ByteBuffer byteBuffer, Callback callback)
+        {
+            WriteCallback writeCallback = new WriteCallback(callback);
+            stateInfo.writeCallback = writeCallback;
```

Comment:
This also replaces the `writeCallback` when a write is done from within a `succeeded()` callback.